### ISSUE TITLE
GroundingDINO Model Error on GPU: Failure to Annotate with GPU Mode

### ISSUE THREAD

**Issue by Kenny-crow** [CVHub520/X-AnyLabeling#389]
大佬您好，有个bug麻烦您过目。
我在使用GoundingDINO（SwinB-QInt8）IDEA这个模型进行人的头部自动标注，在`__preferred_device__ = "CPU"`情况下的结果，可以标注且不会报错。

我已经配置好了GPU运行的环境，并且在DAMO_YOLO等其它模型上也可以正常用GPU运行，但在GroundingDINO上使用GPU运行时，会出现以下报错：

在加载模型时：
` [ 0 ; 9 3 m 2 0 2 4 - 0 4 - 2 5   0 9 : 1 9 : 0 9 . 4 4 8 6 1 0 8   [ W : o n n x r u n t i m e : ,   t r a n s f o r m e r _ m e m c p y . c c : 7 4   o n n x r u n t i m e : : M e m c p y T r a n s f o r m e r : : A p p l y I m p l ]   1 2 6 4   M e m c p y   n o d e s   a r e   a d d e d   t o   t h e   g r a p h   t o r c h _ j i t   f o r   C U D A E x e c u t i o n P r o v i d e r .   I t   m i g h t   h a v e   n e g a t i v e   i m p a c t   o n   p e r f o r m a n c e   ( i n c l u d i n g   u n a b l e   t o   r u n   C U D A   g r a p h ) .   S e t   s e s s i o n _ o p t i o n s . l o g _ s e v e r i t y _ l e v e l = 1   t o   s e e   t h e   d e t a i l   l o g s   b e f o r e   t h i s   m e s s a g e .  [ m 
  [ 0 ; 9 3 m 2 0 2 4 - 0 4 - 2 5   0 9 : 1 9 : 0 9 . 6 4 0 2 0 1 8   [ W : o n n x r u n t i m e : ,   s e s s i o n _ s t a t e . c c : 1 1 6 6   o n n x r u n t i m e : : V e r i f y E a c h N o d e I s A s s i g n e d T o A n E p ]   S o m e   n o d e s   w e r e   n o t   a s s i g n e d   t o   t h e   p r e f e r r e d   e x e c u t i o n   p r o v i d e r s   w h i c h   m a y   o r   m a y   n o t   h a v e   a n   n e g a t i v e   i m p a c t   o n   p e r f o r m a n c e .   e . g .   O R T   e x p l i c i t l y   a s s i g n s   s h a p e   r e l a t e d   o p s   t o   C P U   t o   i m p r o v e   p e r f .  [ m 
  [ 0 ; 9 3 m 2 0 2 4 - 0 4 - 2 5   0 9 : 1 9 : 0 9 . 6 4 0 5 9 2 9   [ W : o n n x r u n t i m e : ,   s e s s i o n _ s t a t e . c c : 1 1 6 8   o n n x r u n t i m e : : V e r i f y E a c h N o d e I s A s s i g n e d T o A n E p ]   R e r u n n i n g   w i t h   v e r b o s e   o u t p u t   o n   a   n o n - m i n i m a l   b u i l d   w i l l   s h o w   n o d e   a s s i g n m e n t s .  [ m 
 `
在自动标注时：
`Error in predict_shapes: [ONNXRuntimeError] : 1 : FAIL : Non-zero status code returned while running Expand node. Name:'Expand_33526' Status Message: Expand_33526: left operand cannot broadcast on dim 2 LeftShape: {1,900,4}, RightShape: {1,900,256}`

从而无法正常标注。我没有找到解决该问题的有效途径，恳请指点

**Comment by Hermione-ff** [comment]
我也出现了类似的问题，请问您解决了吗

**Comment by Hermione-ff** [comment]
> 大佬您好，有个bug麻烦您过目。 我在使用GoundingDINO（SwinB-QInt8）IDEA这个模型进行人的头部自动标注，在`__preferred_device__ = "CPU"`情况下的结果，可以标注且不会报错。
> 
> 我已经配置好了GPU运行的环境，并且在DAMO_YOLO等其它模型上也可以正常用GPU运行，但在GroundingDINO上使用GPU运行时，会出现以下报错：
> 
> 在加载模型时： `� [ 0 ; 9 3 m 2 0 2 4 - 0 4 - 2 5 0 9 : 1 9 : 0 9 . 4 4 8 6 1 0 8 [ W : o n n x r u n t i m e : , t r a n s f o r m e r _ m e m c p y . c c : 7 4 o n n x r u n t i m e : : M e m c p y T r a n s f o r m e r : : A p p l y I m p l ] 1 2 6 4 M e m c p y n o d e s a r e a d d e d t o t h e g r a p h t o r c h _ j i t f o r C U D A E x e c u t i o n P r o v i d e r . I t m i g h t h a v e n e g a t i v e i m p a c t o n p e r f o r m a n c e ( i n c l u d i n g u n a b l e t o r u n C U D A g r a p h ) . S e t s e s s i o n _ o p t i o n s . l o g _ s e v e r i t y _ l e v e l = 1 t o s e e t h e d e t a i l l o g s b e f o r e t h i s m e s s a g e . � [ m � [ 0 ; 9 3 m 2 0 2 4 - 0 4 - 2 5 0 9 : 1 9 : 0 9 . 6 4 0 2 0 1 8 [ W : o n n x r u n t i m e : , s e s s i o n _ s t a t e . c c : 1 1 6 6 o n n x r u n t i m e : : V e r i f y E a c h N o d e I s A s s i g n e d T o A n E p ] S o m e n o d e s w e r e n o t a s s i g n e d t o t h e p r e f e r r e d e x e c u t i o n p r o v i d e r s w h i c h m a y o r m a y n o t h a v e a n n e g a t i v e i m p a c t o n p e r f o r m a n c e . e . g . O R T e x p l i c i t l y a s s i g n s s h a p e r e l a t e d o p s t o C P U t o i m p r o v e p e r f . � [ m � [ 0 ; 9 3 m 2 0 2 4 - 0 4 - 2 5 0 9 : 1 9 : 0 9 . 6 4 0 5 9 2 9 [ W : o n n x r u n t i m e : , s e s s i o n _ s t a t e . c c : 1 1 6 8 o n n x r u n t i m e : : V e r i f y E a c h N o d e I s A s s i g n e d T o A n E p ] R e r u n n i n g w i t h v e r b o s e o u t p u t o n a n o n - m i n i m a l b u i l d w i l l s h o w n o d e a s s i g n m e n t s . � [ m ` 在自动标注时： `Error in predict_shapes: [ONNXRuntimeError] : 1 : FAIL : Non-zero status code returned while running Expand node. Name:'Expand_33526' Status Message: Expand_33526: left operand cannot broadcast on dim 2 LeftShape: {1,900,4}, RightShape: {1,900,256}`
> 
> 从而无法正常标注。我没有找到解决该问题的有效途径，恳请指点

我也出现了类似的问题，请问您解决了吗

**Comment by Hermione-ff** [comment]
> 大佬您好，有个bug麻烦您过目。 我在使用GoundingDINO（SwinB-QInt8）IDEA这个模型进行人的头部自动标注，在`__preferred_device__ = "CPU"`情况下的结果，可以标注且不会报错。
> 
> 我已经配置好了GPU运行的环境，并且在DAMO_YOLO等其它模型上也可以正常用GPU运行，但在GroundingDINO上使用GPU运行时，会出现以下报错：
> 
> 在加载模型时： `� [ 0 ; 9 3 m 2 0 2 4 - 0 4 - 2 5 0 9 : 1 9 : 0 9 . 4 4 8 6 1 0 8 [ W : o n n x r u n t i m e : , t r a n s f o r m e r _ m e m c p y . c c : 7 4 o n n x r u n t i m e : : M e m c p y T r a n s f o r m e r : : A p p l y I m p l ] 1 2 6 4 M e m c p y n o d e s a r e a d d e d t o t h e g r a p h t o r c h _ j i t f o r C U D A E x e c u t i o n P r o v i d e r . I t m i g h t h a v e n e g a t i v e i m p a c t o n p e r f o r m a n c e ( i n c l u d i n g u n a b l e t o r u n C U D A g r a p h ) . S e t s e s s i o n _ o p t i o n s . l o g _ s e v e r i t y _ l e v e l = 1 t o s e e t h e d e t a i l l o g s b e f o r e t h i s m e s s a g e . � [ m � [ 0 ; 9 3 m 2 0 2 4 - 0 4 - 2 5 0 9 : 1 9 : 0 9 . 6 4 0 2 0 1 8 [ W : o n n x r u n t i m e : , s e s s i o n _ s t a t e . c c : 1 1 6 6 o n n x r u n t i m e : : V e r i f y E a c h N o d e I s A s s i g n e d T o A n E p ] S o m e n o d e s w e r e n o t a s s i g n e d t o t h e p r e f e r r e d e x e c u t i o n p r o v i d e r s w h i c h m a y o r m a y n o t h a v e a n n e g a t i v e i m p a c t o n p e r f o r m a n c e . e . g . O R T e x p l i c i t l y a s s i g n s s h a p e r e l a t e d o p s t o C P U t o i m p r o v e p e r f . � [ m � [ 0 ; 9 3 m 2 0 2 4 - 0 4 - 2 5 0 9 : 1 9 : 0 9 . 6 4 0 5 9 2 9 [ W : o n n x r u n t i m e : , s e s s i o n _ s t a t e . c c : 1 1 6 8 o n n x r u n t i m e : : V e r i f y E a c h N o d e I s A s s i g n e d T o A n E p ] R e r u n n i n g w i t h v e r b o s e o u t p u t o n a n o n - m i n i m a l b u i l d w i l l s h o w n o d e a s s i g n m e n t s . � [ m ` 在自动标注时： `Error in predict_shapes: [ONNXRuntimeError] : 1 : FAIL : Non-zero status code returned while running Expand node. Name:'Expand_33526' Status Message: Expand_33526: left operand cannot broadcast on dim 2 LeftShape: {1,900,4}, RightShape: {1,900,256}`
> 
> 从而无法正常标注。我没有找到解决该问题的有效途径，恳请指点

我把我的onnxruntime-gpu从1.17.1降到1.16.0解决了这个问题，要根据你的cuda和cudnn版本来安装onnxruntime-gpu，详情参考https://onnxruntime.ai/docs/execution-providers/CUDA-ExecutionProvider.html

**Comment by Justin900429** [comment]
我也有遇到類似的問題。目前猜測是sequence length 的問題，解法的話是透過 padding 到 max_length。

首先替換掉 `tokenizer`:

```python
@staticmethod
def get_tokenlizer(text_encoder_type="bert-base-uncased"):
    from transformers import BertTokenizer
    tokenizer = BertTokenizer.from_pretrained(text_encoder_type)
    return tokenizer
```

接著將 `preprocess` 換成:

```python
def preprocess(self, image, text_prompt, img_mask=None):
        # encoder texts
        captions = self.get_caption(str(text_prompt))
        tokenized_raw_results = self.net.tokenizer([captions], padding="max_length", max_length=256, return_tensors="np")
        
        tokenized = {
            "input_ids": tokenized_raw_results.input_ids.astype(np.int64),
            "token_type_ids": tokenized_raw_results.token_type_ids.astype(np.int64),
            "attention_mask": tokenized_raw_results.attention_mask,
        }
        specical_tokens = [101, 102, 1012, 1029]
        (
            text_self_attention_masks,
            position_ids,
            _,
        ) = self.generate_masks_with_special_tokens_and_transfer_map(tokenized, specical_tokens)
        if text_self_attention_masks.shape[1] > self.net.max_text_len:
            text_self_attention_masks = text_self_attention_masks[
                :, : self.net.max_text_len, : self.net.max_text_len
            ]

            position_ids = position_ids[:, : self.net.max_text_len]
            tokenized["input_ids"] = tokenized["input_ids"][:, : self.net.max_text_len]
            tokenized["attention_mask"] = tokenized["attention_mask"][:, : self.net.max_text_len]
            tokenized["token_type_ids"] = tokenized["token_type_ids"][:, : self.net.max_text_len]
        inputs = {}
        inputs["img"] = image
        inputs["input_ids"] = np.array(tokenized["input_ids"], dtype=np.int64)
        inputs["attention_mask"] = np.array(tokenized["attention_mask"], dtype=bool)
        inputs["token_type_ids"] = np.array(tokenized["token_type_ids"], dtype=np.int64)
        inputs["position_ids"] = np.array(position_ids, dtype=np.int64)
        inputs["text_token_mask"] = np.array(text_self_attention_masks, dtype=bool)
        return image, inputs, captions
```

最後將 `postprocess`  換成

```python
def postprocess(self, outputs, caption, with_logits=True, token_spans=None):
        logits, boxes = outputs
        logits_filt = np.squeeze(logits, 0)  # [0]  # prediction_logits.shape = (nq, 256)
        logits_filt = self.sig(logits_filt)
        boxes_filt = np.squeeze(boxes, 0)  # [0]  # prediction_boxes.shape = (nq, 4)
        # filter output
        if token_spans is None:
            filt_mask = logits_filt.max(axis=1) > self.box_threshold
            logits_filt = logits_filt[filt_mask]  # num_filt, 256
            boxes_filt = boxes_filt[filt_mask]  # num_filt, 4

            # get phrase
            tokenlizer = self.net.tokenizer
            tokenized_raw_results = tokenlizer(caption, return_tensors="np")
            tokenized = {
                "input_ids": tokenized_raw_results.input_ids[0].astype(np.int64),
                "token_type_ids": tokenized_raw_results.token_type_ids[0].astype(np.int64),
                "attention_mask": tokenized_raw_results.attention_mask[0],
            }
            # build pred
            pred_phrases = []
            for logit in logits_filt:
                posmap = logit > self.text_threshold
                pred_phrase = self.get_phrases_from_posmap(posmap, tokenized, tokenlizer)
                if with_logits:
                    pred_phrases.append([pred_phrase.strip(), logit.max()])
                else:
                    pred_phrases.append([pred_phrase.strip(), 1.0])
        else:
            # TODO: Using token_spans.
            raise NotImplementedError
        return boxes_filt, pred_phrases
```

這樣應該可以解決問題。`postprocess` 最後不用 padding 是因為這部只需要拿到 label 而已，因此不會出錯。
